File: app/static/versions.mjs
// T066 (US6, UX-AC06): the owner's operating versions. What runs now and what ran
// before; each candidate with the gates its validation actually recorded (failed and
// invalid gates keep their reasons); approve / reject / defer as the owner's own
// decision — approve is offered only for a passed sealed-offline candidate — and
// applying an approved candidate as a separate explicit act on the exact revision
// shown; a rollback that requires a stated reason and never claims to undo what
// already happened outside. Growth experiments show the stop reason the loop
// recorded, never a reconstructed one. All text reaches the DOM through textContent.

// The isolation boundaries a comparison plan needs (G-14) are listed with what each
// means and are approved or rejected here as the owner's own recorded decision over
// the exact boundary shown (its digest travels with the command).

import { renderBoundaries, renderRounds } from './experiments.mjs';

const BASE_PATH = /^\/(?:[0-9a-f]{32}\/)?$/;

export const STOP_REASONS = Object.freeze({
  plateau_reached: '하한 도달 뒤 세 번 연속 의미 있는 개선이 없어 멈춤',
  budget_exhausted: '하한 도달 뒤 탐색 예산을 모두 씀(수렴이 아님)',
  below_floor_exhausted: '하한에 한 번도 닿지 못한 채 예산을 모두 씀(품질 미달, 승격 아님)',
  evidence_blocked: '필요한 증거가 없어 다음 비교를 할 수 없음',
  safety_stop: '안전 경계 위반으로 멈춤(최고 점수도 승격 근거가 아님)',
  human_stop: '사용자가 멈춤(운영 적용 동의가 아님)',
  lineage_changed: '비교 조건이 바뀌어 새 계보로 다시 해야 함',
});

export const GATE_LABELS = Object.freeze({
  reconstruction: '재구성', heldout_transfer: '미관측 전이', boundary_exclusion: '경계 제외',
  regression: '회귀', leakage: '누출', side_effects: '부수 효과',
});

export const MESSAGES = Object.freeze({
  noState: '운영 버전이 아직 채택되지 않았습니다. 설계 승인으로 준비된 환경을 채택하면 여기에서 버전을 관리합니다.',
  noCandidates: '검증을 마친 후보가 없습니다.',
  noExperiments: '기록된 성장 실험이 없습니다.',
  notApprovable: '봉인 검증을 통과하지 못한 후보는 승인할 수 없습니다. 거절하거나 보류할 수 있습니다.',
  rollbackReason: '되돌리는 이유를 적어 주세요.',
  rollbackNote: '되돌리면 이전 버전의 번들로 돌아갈 뿐, 이미 보낸 것·게시한 것은 되돌리지 않습니다.',
  unreadable: '이 후보의 기록을 정확히 다시 읽지 못해 결정 대상으로 보이지 않습니다.',
});

export const ERROR_MESSAGES = Object.freeze({
  invalid_input: '요청 형식이 맞지 않습니다.',
  unauthenticated: '세션이 없습니다. 시작 화면(./)에서 로그인해 주세요.',
  access_denied: '이 요청은 허용되지 않았습니다.',
  not_found: '해당 후보를 찾지 못했습니다.',
  conflict: '운영 버전이 그사이 바뀌었거나 이 승인은 이미 적용되었습니다. 다시 확인해 주세요.',
  unavailable: '처리하지 못했습니다.',
});

export const BOUNDARY_ERRORS = Object.freeze({
  ...ERROR_MESSAGES,
  not_found: '해당 비교 계획이나 경계를 찾지 못했습니다.',
  conflict: '이 경계의 내용이 화면에 보인 것과 다르거나, 같은 명령이 다른 결정으로 이미 기록되었습니다. 다시 확인해 주세요.',
});

function fail(message) {
  throw new Error(message);
}

// the budget the loop recorded as actually consumed, invalid rounds included
export const budgetText = budget => {
  const entries = budget && typeof budget === 'object' ? Object.entries(budget) : [];
  return entries.length ? entries.map(([name, amount]) => `${name} ${amount}`).join(', ') : '기록 없음';
};

const short = ref =>(ref && typeof ref.id === 'string' ? `${ref.kind} ${ref.id.slice(0, 8)} (${ref.sha256.slice(0, 12)})` : '없음');

export function createVersionsPanel({ root, document, request, basePath = '/', crypto } = {}) {
  if (typeof root?.replaceChildren !== 'function') fail('a root is required');
  if (typeof request !== 'function') fail('a request adapter is required');
  if (typeof crypto?.randomUUID !== 'function') fail('a crypto with randomUUID is required');
  if (typeof basePath !== 'string' || !BASE_PATH.test(basePath)) fail('base path is not a deployment base path');
  const path = `${basePath.slice(0, -1)}/api/v1/versions`;
  let view = null;
  const approvals = new Map();  // validation report sha → approval ref the owner just recorded

  function element(tag, text, attributes = {}) {
    const node = document.createElement(tag);
    if (text !== undefined) node.textContent = text;
    for (const [name, value] of Object.entries(attributes)) node.setAttribute(name, value);
    return node;
  }

  const status = element('p', '', { role: 'status', 'aria-live': 'polite' });
  const current = element('section', undefined, { 'aria-label': '운영 버전' });
  const candidates = element('section', undefined, { 'aria-label': '후보' });
  const experiments = element('section', undefined, { 'aria-label': '성장 실험' });
  const boundaries = element('section', undefined, { 'aria-label': '도구 효과 경계' });
  const rounds = element('section', undefined, { 'aria-label': '비교 라운드' });
  root.replaceChildren(element('h2', '버전'), status, current, candidates, experiments, boundaries, rounds);
  let plans = [];

  function say(text, state) {
    status.textContent = text;
    status.dataset.state = state;
  }

  function refusal(error) {
    const code = Object.hasOwn(ERROR_MESSAGES, error?.code) ? error.code : 'unavailable';
    say(ERROR_MESSAGES[code], code);
  }

  async function command(name, body) {
    try {
      const next = await request(`${path}/${name}`, { method: 'POST', body });
      return next;
    } catch (error) {
      refusal(error);
      throw error;
    }
  }

  function renderCurrent() {
    if (view.state === null) {
      current.replaceChildren(element('h3', '운영 버전'), element('p', MESSAGES.noState));
      return;
    }
    const state = view.state;
    const history = element('ol', undefined, { 'aria-label': '이전 버전' });
    for (const item of state.history) {
      history.append(element('li', `${short(item.environment_ref)} · ${item.lifecycle === 'retired' ? '이전 운영' : '되돌림으로 내림'}`));
    }
    const parts = [element('h3', '운영 버전'), element('p', `지금 운영: ${short(state.current_environment_ref)} · 수정본 ${state.revision}`),
      element('h4', '이력'), state.history.length ? history : element('p', '이전 버전이 없습니다.')];
    if (state.history.some(item => item.lifecycle === 'retired')) {
      const reason = element('textarea', undefined, { id: 'versions-rollback-reason', rows: '2' });
      const go = element('button', '이전 버전으로 되돌리기', { type: 'button' });
      go.addEventListener('click', () => rollback(reason).catch(() => {}));
      parts.push(element('label', MESSAGES.rollbackReason, { for: 'versions-rollback-reason' }), reason,
        element('p', MESSAGES.rollbackNote), go);
    }
    current.replaceChildren(...parts);
  }

  function renderCandidates() {
    const parts = [element('h3', '후보')];
    if (!view.candidates.length) parts.push(element('p', MESSAGES.noCandidates));
    for (const item of view.candidates) {
      const entry = element('article', undefined, { 'aria-label': '후보' });
      if (!item.readable) {
        entry.append(element('p', MESSAGES.unreadable));
        parts.push(entry);
        continue;
      }
      entry.append(element('h4', `${short(item.candidate_bundle_ref)} · 검증 ${item.status} (${item.mode})`));
      const gates = element('ul', undefined, { 'aria-label': '검증 게이트' });
      for (const [name, gate] of Object.entries(item.gates)) {
        const reasons = gate.reasons.length ? ` · ${gate.reasons.join(', ')}` : '';
        gates.append(element('li', `${GATE_LABELS[name] ?? name}: ${gate.status} · 비교 라운드 ${gate.rounds}개${reasons}`));
      }
      entry.append(gates);
      if (!item.approvable) entry.append(element('p', MESSAGES.notApprovable));
      const decisions = [['reject', '거절'], ['defer', '보류']];
      if (item.approvable) decisions.unshift(['approve', '승인']);
      for (const [decision, label] of decisions) {
        const button = element('button', label, { type: 'button' });
        button.addEventListener('click', () => decide(item, decision).catch(() => {}));
        entry.append(button);
      }
      const approval = approvals.get(item.validation_report_ref.sha256);
      if (approval && view.state !== null) {
        const apply = element('button', '승인한 이 버전 적용', { type: 'button' });
        apply.addEventListener('click', () => activate(approval).catch(() => {}));
        entry.append(apply);
      }
      parts.push(entry);
    }
    candidates.replaceChildren(...parts);
  }

  function renderExperiments() {
    const parts = [element('h3', '성장 실험')];
    if (!view.experiments.length) parts.push(element('p', MESSAGES.noExperiments));
    for (const item of view.experiments) {
      if (!item.readable) {
        parts.push(element('p', `계보 ${item.lineage_id.slice(0, 8)}: 기록을 정확히 다시 읽지 못했습니다.`));
        continue;
      }
      const stop = item.stop_reason ? (STOP_REASONS[item.stop_reason] ?? item.stop_reason) : '진행 중(멈춤 사유 없음)';
      const best = item.best_observed ? `최고 ${item.best_observed.utility} (${item.best_observed.round_id})` : '유효한 최고 기록 없음';
      parts.push(element('p', `계보 ${item.lineage_id.slice(0, 8)} · ${item.status} · ${stop} · ${best} · `
        + `완료 라운드 ${item.completed_round_ids.length}개 · 연속 비개선 ${item.non_improving_valid_count} · `
        + `소비 ${budgetText(item.consumed_budget)}`));
    }
    experiments.replaceChildren(...parts);
  }

  function render(next) {
    view = next;
    renderCurrent();
    renderCandidates();
    renderExperiments();
    renderRounds({ root: rounds, document, rounds: Array.isArray(view.rounds) ? view.rounds : [] });
  }

  function renderPlans() {
    renderBoundaries({ root: boundaries, document, plans, onDecide: decideBoundary });
  }

  async function loadBoundaries() {
    try {
      const answer = await request(`${path}/tool-effect-boundaries`, {});
      plans = Array.isArray(answer?.plans) ? answer.plans : null;
    } catch {
      plans = null;  // said plainly in the section; the rest of the page still loads
    }
    renderPlans();
    return plans;
  }

  async function load() {
    try {
      const next = await request(path, {});
      render(next);
      await loadBoundaries();
      say('', 'loaded');
      return next;
    } catch (error) {
      refusal(error);
      throw error;
    }
  }

  async function decideBoundary(plan, item, decision) {
    let answer;
    try {
      answer = await request(`${path}/tool-effect-boundaries/decisions`, { method: 'POST', body: {
        command_id: crypto.randomUUID(), plan_record_ref: plan.plan_record, tool_id: item.tool_id,
        version: item.version, boundary_sha256: item.boundary_sha256, decision } });
    } catch (error) {
      const code = Object.hasOwn(BOUNDARY_ERRORS, error?.code) ? error.code : 'unavailable';
      say(BOUNDARY_ERRORS[code], code);
      throw error;
    }
    const updated = answer?.boundary;
    if (updated && Array.isArray(plans)) {
      plans = plans.map(entry => (entry.plan_record?.sha256 !== plan.plan_record.sha256 ? entry : {
        ...entry, boundaries: entry.boundaries.map(value => (value.tool_id === updated.tool_id
          && value.version === updated.version ? updated : value)) }));
    }
    renderPlans();
    say(decision === 'approve' ? '경계 승인을 기록했습니다. 이 경계만 쓰며 실제 서비스로는 보내지 않습니다.'
      : '경계 거절을 기록했습니다. 이 경계가 필요한 항목은 비교하지 않습니다.', 'boundary_decided');
    return answer;
  }

  async function decide(item, decision) {
    const answer = await command('decisions', { command_id: crypto.randomUUID(), decision,
      validation_report_ref: item.validation_report_ref });
    if (decision === 'approve') approvals.set(item.validation_report_ref.sha256, answer.approval_ref);
    say(decision === 'approve' ? '승인을 기록했습니다. 적용은 따로 합니다.' : decision === 'reject' ? '거절을 기록했습니다.'
      : '보류를 기록했습니다.', 'decided');
    renderCandidates();
    return answer;
  }

  async function activate(approvalRef) {
    const next = await command('activate', { approval_ref: approvalRef, expected_revision: view.state.revision });
    render(next);
    say('승인한 버전을 운영 버전으로 적용했습니다.', 'applied');
    return next;
  }

  async function rollback(reasonField) {
    const reason = String(reasonField.value ?? '').trim();
    if (!reason) {
      say(MESSAGES.rollbackReason, 'invalid_input');
      return null;
    }
    const next = await command('rollback', { reason, expected_revision: view.state.revision });
    render(next);
    say('이전 버전으로 되돌렸습니다. 외부로 이미 나간 효과는 되돌리지 않았습니다.', 'rolled_back');
    return next;
  }

  return Object.freeze({ load, decide, decideBoundary, activate, get view() { return view; },
    get plans() { return plans; } });
}

File: app/services/versions.py
"""The owner's operating versions: what runs now, what is waiting, approve, apply,
roll back (US6, T066 server half; growth.md §8, FR-026, UX-AC06).

One operating scope per vault holds a `PromotionState` lineage (growth_store's
promotion records, CAS by revision). The owner adopts the current environment once;
after that the only ways the operating version changes are an exact approved
candidate applied through `activate_candidate` (the approval binds the candidate
bundle, the validation report the owner saw and the current environment it was
given against; an approval is consumed once) or an explicit rollback with a stated
reason to the last retired version. Candidates are the persisted validation
reports whose bundle and cited rounds resume exactly (growth_store); a candidate
without a passed sealed-offline report can be rejected or deferred, never
approved. Nothing here runs a model, a tool or an external effect, and a rollback
restores the bundle, never the world (`external_effects_reverted` stays false).
"""

from __future__ import annotations

from functools import wraps
from uuid import NAMESPACE_URL, uuid5

from ..domain.refs import EntityRef, uuid_string
from ..domain.store import DomainStore
from .growth_store import (
    RECORD_KIND,
    GrowthStoreError,
    persist_promotion_state,
    resume_comparison_round_record,
    resume_loop,
    resume_promotion_state,
    resume_validation_report,
)
from .owner_auth import OwnerAuthError
from .promotion import (
    PromotionError,
    activate_candidate,
    open_promotion_state,
    record_promotion_decision,
    rollback_environment,
)
from .promotion_approvals import PersistentPromotionApprovals, PromotionApprovalError
from .run_approvals import _authenticate_owner, _owner_actor_ref
from .validation import validation_report_ref

__all__ = ["PersistentVersions", "VersionError"]

CODES = frozenset({"invalid_input", "unauthenticated", "access_denied", "not_found", "conflict",
                   "unavailable"})
MAX_CANDIDATES = 64
MAX_ROUNDS = 256
_STAMP_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"


class VersionError(ValueError):
    def __init__(self, code="invalid_input"):
        if code not in CODES:
            code = "unavailable"
        super().__init__(code)
        self.code = code


def _closed(method):
    @wraps(method)
    def invoke(*args, **kwargs):
        try:
            return method(*args, **kwargs)
        except (VersionError, OwnerAuthError):
            raise
        except PromotionApprovalError as error:
            raise VersionError("conflict" if str(error) == "conflict" else "invalid_input") from None
        except (PromotionError, GrowthStoreError):
            raise VersionError("conflict") from None
        except Exception:  # noqa: BLE001 - storage errors must not disclose detail
            raise VersionError("unavailable") from None

    return invoke


def _stamp():
    from datetime import UTC, datetime

    return datetime.now(UTC).strftime(_STAMP_FORMAT)


class PersistentVersions:
    def __init__(self, domain_store, owner_authority, approvals):
        if type(domain_store) is not DomainStore:
            raise TypeError("Exact DomainStore required")
        if type(approvals) is not PersistentPromotionApprovals:
            raise TypeError("Exact PersistentPromotionApprovals required")
        self._domain = domain_store
        self._owner = owner_authority
        self._approvals = approvals

    # --- the scope's promotion lineage -----------------------------------------------

    def _scope(self, roots) -> str:
        return str(uuid5(NAMESPACE_URL, f"deeptwin:operating-scope:{roots.genesis.id}"))

    def _head(self):
        """(state, record ref) of the newest promotion revision, or (None, None)."""

        with self._domain._connection() as db:
            roots = self._domain._read_roots(db)
            scope = self._scope(roots)
            record_id = str(uuid5(NAMESPACE_URL, f"deeptwin:promotion:{scope}"))
            row = db.execute(
                "SELECT version, sha256 FROM domain_records WHERE vault_id=? AND kind=? AND id=? "
                "ORDER BY version DESC LIMIT 1", (roots.genesis.id, RECORD_KIND, record_id)).fetchone()
        if row is None:
            return None, None, scope, roots
        ref = EntityRef(RECORD_KIND, record_id, row["version"], row["sha256"])
        return resume_promotion_state(self._domain, ref), ref, scope, roots

    def _headers(self, request, roots):
        with self._domain._connection() as db:
            actor = _authenticate_owner(self._owner, request)
            actor_ref = _owner_actor_ref(db, actor)
        return {"actor_ref": actor_ref, "access_policy_ref": roots.access_policy,
                "retention_policy_ref": roots.retention_policy, "created_at_utc": _stamp()}

    def _candidates(self):
        with self._domain._connection() as db:
            roots = self._domain._read_roots(db)
            rows = db.execute(
                "SELECT id, version, sha256 FROM domain_records WHERE vault_id=? AND kind=? "
                "AND instr(body, ?) > 0 ORDER BY id LIMIT ?",
                (roots.genesis.id, RECORD_KIND, b'"growth_kind":"growth_validation_report"',
                 MAX_CANDIDATES)).fetchall()
        found = []
        for row in rows:
            ref = EntityRef(RECORD_KIND, row["id"], row["version"], row["sha256"])
            try:
                candidate, report = resume_validation_report(self._domain, ref)
            except GrowthStoreError:
                # a report whose chain does not resume is shown as such, never as a candidate
                found.append({"report_record": ref.as_dict(), "readable": False})
                continue
            found.append({
                "report_record": ref.as_dict(), "readable": True,
                "validation_report_ref": validation_report_ref(report).as_dict(),
                "candidate_bundle_ref": candidate.bundle_ref.as_dict(),
                "change_candidate_ref": candidate.candidate.as_dict(),
                "mode": report.mode, "status": report.status,
                "approvable": report.mode == "sealed_offline" and report.status == "passed",
                "gates": {name: {"status": outcome.status, "reasons": list(outcome.reasons),
                                 "rounds": len(outcome.evidence)} for name, outcome in report.gates},
            })
        return found

    def _experiments(self):
        """Each growth lineage's newest loop revision, resumed exactly: status, the stop
        reason the loop actually recorded, best and progress rounds and counters."""

        with self._domain._connection() as db:
            roots = self._domain._read_roots(db)
            rows = db.execute(
                "SELECT id, MAX(version) AS version FROM domain_records WHERE vault_id=? AND kind=? "
                "AND instr(body, ?) > 0 GROUP BY id ORDER BY id LIMIT ?",
                (roots.genesis.id, RECORD_KIND, b'"growth_kind":"growth_loop_state"', MAX_CANDIDATES)).fetchall()
            refs = []
            for row in rows:
                digest = db.execute("SELECT sha256 FROM domain_records WHERE vault_id=? AND kind=? AND id=? "
                                    "AND version=?", (roots.genesis.id, RECORD_KIND, row["id"], row["version"])).fetchone()
                refs.append(EntityRef(RECORD_KIND, row["id"], row["version"], digest["sha256"]))
        found = []
        for ref in refs:
            try:
                loop = resume_loop(self._domain, ref).as_dict()
            except GrowthStoreError:
                found.append({"lineage_id": ref.id, "readable": False})
                continue
            found.append({"readable": True, **{key: loop[key] for key in (
                "lineage_id", "revision", "status", "floor_reached", "best_observed", "progress_reference",
                "non_improving_valid_count", "completed_round_ids", "consumed_budget", "stop_reason")}})
        return found

    def _rounds(self):
        """Every persisted paired round, re-issued exactly over its stored plan: which
        baseline run is paired with which candidate run, the round's validity and its
        stated reasons, and — only on a valid round — the measurements and utility it
        recorded. Nothing is scored here; a round that does not read back is listed as
        unreadable, never dropped."""

        with self._domain._connection() as db:
            roots = self._domain._read_roots(db)
            rows = db.execute(
                "SELECT id, version, sha256 FROM domain_records WHERE vault_id=? AND kind=? AND version=1 "
                "AND instr(body, ?) > 0 ORDER BY id LIMIT ?",
                (roots.genesis.id, RECORD_KIND, b'"growth_kind":"growth_comparison_round"',
                 MAX_ROUNDS)).fetchall()
        found = []
        for row in rows:
            ref = EntityRef(RECORD_KIND, row["id"], row["version"], row["sha256"])
            try:
                plan, result = resume_comparison_round_record(self._domain, ref)
            except GrowthStoreError:
                found.append({"round_record": ref.as_dict(), "readable": False})
                continue
            value = result.as_dict()
            found.append({
                "round_record": ref.as_dict(), "readable": True, "lineage_id": plan.lineage_id,
                "plan_mode": plan.mode, "baseline_environment_ref": plan.baseline_environment.as_dict(),
                "round_id": value["round_id"], "round_index": value["round_index"],
                "candidate_ref": value["candidate_ref"],
                "pairs": [{"baseline_run_ref": baseline, "candidate_run_ref": candidate}
                          for baseline, candidate in zip(value["baseline_run_refs"],
                                                         value["candidate_run_refs"], strict=True)],
                "validity": value["validity"], "validity_reasons": value["validity_reasons"],
                "metric_vector": value["metric_vector"], "utility": value["utility"],
                "evidence_refs": value["evidence_refs"],
            })
        found.sort(key=lambda item: (item.get("lineage_id", ""), item.get("round_index", -1),
                                     item["round_record"]["id"]))
        return found

    @staticmethod
    def _state_view(state):
        if state is None:
            return None
        return {"current_environment_ref": state.current_environment.as_dict(),
                "history": [{"environment_ref": ref.as_dict(), "lifecycle": lifecycle}
                            for ref, lifecycle in state.history],
                "applied_approvals": len(state.consumed_decisions),
                "external_effects_reverted": state.external_effects_reverted,
                "revision": state.revision}

    # --- reads --------------------------------------------------------------------------

    @_closed
    def read(self, request) -> dict:
        if request is None:
            raise VersionError("unauthenticated")
        self._owner.authenticate_bound(request.session)
        state, _ref, _scope, _roots = self._head()
        return {"state": self._state_view(state), "candidates": self._candidates(),
                "experiments": self._experiments(), "rounds": self._rounds()}

    # --- commands -----------------------------------------------------------------------

    @_closed
    def adopt(self, request, payload) -> dict:
        """Adopt an existing environment record as the first operating version, once."""

        if type(payload) is not dict or set(payload) != {"environment_ref"}:
            raise VersionError("invalid_input")
        _authenticate_owner(self._owner, request)
        environment = EntityRef.from_dict(payload["environment_ref"])
        if environment.kind != "environment":
            raise VersionError("invalid_input")
        self._domain.get(environment)  # it must be an actual record of this vault
        state, _ref, scope, roots = self._head()
        if state is not None:
            raise VersionError("conflict")  # adoption happens once; later changes are promotions
        persist_promotion_state(self._domain, open_promotion_state(environment.as_dict()),
                                scope_id=scope, parent_ref=None, **self._headers(request, roots))
        return self.read(request)

    @_closed
    def decide(self, request, payload) -> dict:
        """Record the owner's approve / reject / defer over one exact candidate."""

        if type(payload) is not dict or set(payload) != {"command_id", "decision", "validation_report_ref"}:
            raise VersionError("invalid_input")
        uuid_string(payload["command_id"])
        state, _ref, _scope, _roots = self._head()
        if state is None:
            raise VersionError("conflict")  # nothing operates yet: adopt first
        candidate = self._candidate_for(payload["validation_report_ref"])
        approval = self._approvals.record(request, {
            "schema_version": "promotion-approval-command-v1", "command_id": payload["command_id"],
            "decision": payload["decision"],
            "candidate_bundle": candidate["candidate_bundle_ref"],
            "validation_report": candidate["validation_report_ref"],
            "expected_current_environment": state.current_environment.as_dict(),
        })
        return {"approval_ref": approval.approval_ref.as_dict(), "decision": approval.decision,
                "candidate_bundle_ref": candidate["candidate_bundle_ref"]}

    def _candidate_for(self, report_ref_value):
        wanted = EntityRef.from_dict(report_ref_value)
        for item in self._candidates():
            if item["readable"] and item["validation_report_ref"] == wanted.as_dict():
                return item
        raise VersionError("not_found")

    @_closed
    def activate(self, request, payload) -> dict:
        """Apply exactly the approved bundle, or fail honestly (G-11/G-13)."""

        if type(payload) is not dict or set(payload) != {"approval_ref", "expected_revision"}:
            raise VersionError("invalid_input")
        _authenticate_owner(self._owner, request)
        approval = self._approvals.resolve(EntityRef.from_dict(payload["approval_ref"]))
        state, head_ref, scope, roots = self._head()
        if state is None or state.revision != payload["expected_revision"]:
            raise VersionError("conflict")
        if self._domain.get(head_ref).body["created_at_utc"] > approval.decided_at_utc:
            # The operating version moved after the owner decided (G-13). Matching the
            # expected environment is not enough: after an apply and a rollback the same
            # environment is current again, but the owner decided over a state that no
            # longer holds, so the approval is never revived — it is given again.
            raise VersionError("conflict")
        candidate_item = self._candidate_for(approval.validation_report.as_dict())
        candidate, report = resume_validation_report(
            self._domain, EntityRef.from_dict(candidate_item["report_record"]))
        decision = record_promotion_decision({
            "candidate": candidate, "validation_report": report, "scope": candidate.scope.as_dict(),
            "approval": approval, "expected_current_environment": approval.expected_current_environment.as_dict(),
            "rollback_bundle": candidate.rollback_bundle.as_dict(),
        })
        applied = activate_candidate(state, decision, candidate)
        persist_promotion_state(self._domain, applied, scope_id=scope, parent_ref=head_ref,
                                **self._headers(request, roots))
        return self.read(request)

    @_closed
    def rollback(self, request, payload) -> dict:
        """Restore the last retired version, stating why; nothing external is undone."""

        if (type(payload) is not dict or set(payload) != {"reason", "expected_revision"}
                or type(payload["reason"]) is not str):
            raise VersionError("invalid_input")
        _authenticate_owner(self._owner, request)
        state, head_ref, scope, roots = self._head()
        if state is None or state.revision != payload["expected_revision"]:
            raise VersionError("conflict")
        restored = rollback_environment(state, payload["reason"])
        persist_promotion_state(self._domain, restored, scope_id=scope, parent_ref=head_ref,
                                **self._headers(request, roots))
        return self.read(request)
